news/news.1
.\"
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)news.1	1.4 (gritter) 4/17/03
.TH NEWS 1 "4/17/03" "Heirloom Toolchest" "User Commands"
.SH NAME
news \- print news items
.SH SYNOPSIS
\fBnews\fR [\fB\-a\fR] [\fB\-n\fR] [\fB\-s\fR] [\fIitems\fR]
.SH DESCRIPTION
The
.B news
command prints files from the systems news directory
that have been added or modified
since its previous invocation,
preceded by a header,
starting with the most recent file.
.PP
It accepts the following options:
.TP
.B \-a
Print all files in the system news directory.
The recorded time of the previous invocation is not changed.
.TP
.B \-n
Print only the names of added or modified files.
The recorded time of the previous invocation is not changed.
.TP
.B \-s
Print only the count of added or modified files.
The recorded time of the previous invocation is not changed.
.SH "ENVIRONMENT VARIABLES"
.TP
.B HOME
User's home directory.
.SH FILES
.TP
\fB$HOME/.news_time\fR
The time of the previous invocation is recorded
as the modification time of this file.
.TP
\fB/var/news\fR
System news directory.


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

news/news.c
/*
 * news - print news items
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, December 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)news.sl	1.11 (gritter) 5/29/05";

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<dirent.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<time.h>
#include	<libgen.h>
#include	<utime.h>
#include	<pwd.h>
#include	<setjmp.h>
#include	<signal.h>
#include	"sigset.h"

#include	"iblok.h"
#include	"oblok.h"

struct	item {
	const char	*i_name;
	time_t		i_time;
	uid_t		i_uid;
};

static const char	newsdir[] = "/var/news";
static const char	currency[] = ".news_time";

static unsigned		errcnt;		/* count of errors */
static int		aflag;		/* print all items */
static int		nflag;		/* print names of items only */
static int		sflag;		/* print item count */
static char		*progname;	/* argv[0] to main() */
static unsigned long	item_count;	/* count of items */
static time_t		news_time;	/* time of last news printing */
static time_t		start_time;	/* time at start of command */
static int		canjump;	/* jmpenv is valid */
static jmp_buf		jmpenv;		/* jump buffer */
static struct oblok	*output;	/* output buffer */

/*
 * Memory allocation with check.
 */
void *
srealloc(void *vp, size_t nbytes)
{
	void	*p;

	if ((p = (void *)realloc(vp, nbytes)) == NULL) {
		write(2, "No storage\n", 11);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t nbytes)
{
	return srealloc(NULL, nbytes);
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s [-a] [-n] [-s] [items]\n", progname);
	exit(2);
}

static char *
components(const char *prefix, const char *name)
{
	char	*cp;

	if (prefix) {
		if (name) {
			cp = smalloc(strlen(prefix) + strlen(name) + 2);
			strcpy(cp, prefix);
			strcat(cp, "/");
			strcat(cp, name);
		} else {
			cp = smalloc(strlen(prefix) + 1);
			strcpy(cp, prefix);
		}
	} else {
		cp = smalloc(strlen(name) + 1);
		strcpy(cp, name);
	}
	return cp;
}

static void
cantopen(const char *name)
{
	char	*file;

	file = components(newsdir, name);
	fprintf(stderr, "Cannot open %s\n", file);
	free(file);
	errcnt |= 1;
}

static void
onint(int signo)
{
	static time_t	cur, last;

	time(&cur);
	if (cur - last <= 1)
		exit(0200 + signo);
	if (canjump)
		longjmp(jmpenv, signo);
}

static void
cat(int fd)
{
	struct iblok	*ip;
	int	signo;
	int	c, lastc = '\n';

#ifdef	__GNUC__
	(void)&lastc;
#endif	/* __GNUC__ */
	ip = ib_alloc(fd, 0);
	if ((signo = setjmp(jmpenv)) == 0) {
		canjump = 1;
		while ((c = ib_get(ip)) != EOF) {
			if (lastc == '\n')
				ob_write(output, "   ", 3);
			ob_put(c, output);
			lastc = c;
		}
		canjump = 0;
	} else
		sigrelse(signo);
	ib_free(ip);
}

static char *
user(uid_t uid)
{
	struct passwd	*pw;
	char	*cp;
	size_t sz;

	if ((pw = getpwuid(uid)) != NULL) {
		cp = smalloc(sz = strlen(pw->pw_name) + 3);
		snprintf(cp, sz, "(%s)", pw->pw_name);
	} else {
		const char	dots[] = ".....";

		cp = smalloc(strlen(dots) + 1);
		strcpy(cp, dots);
	}
	return cp;
}

static void
item(const struct item *ip)
{
	if (nflag) {
		if (item_count++ == 0)
			printf("news:");
		printf(" %s", ip->i_name);
	} else if (!sflag) {
		char	*cp, *up;
		int	fd, n, sz;

		if ((fd = open(ip->i_name, O_RDONLY)) >= 0) {
			up = user(ip->i_uid);
			cp = malloc(sz = strlen(ip->i_name)+strlen(up)+30);
			n = snprintf(cp, sz, "%s%s %s %s",
				item_count ? "" : "\n",
				ip->i_name, up, ctime(&ip->i_time));
			ob_write(output, cp, n);
			free(cp);
			free(up);
			cat(fd);
			close(fd);
			ob_write(output, "\n", 1);
			item_count++;
		} else
			cantopen(ip->i_name);
	} else
		item_count++;
}

static time_t
gettime(const char *prefix, const char *name, uid_t *uid)
{
	struct stat	st;
	char	*file;
	time_t	t;

	file = components(prefix, name);
	if (stat(file, &st) == 0) {
		if (uid)
			*uid = st.st_uid;
		t = st.st_mtime;
	} else
		t = 0;
	free(file);
	return t;
}

static void
touch(const char *prefix, const char *name)
{
	struct stat	st;
	struct utimbuf	ut;
	char	*file;

	file = components(prefix, name);
	if (stat(file, &st) < 0) {
		close(open(file, O_WRONLY|O_CREAT, 0666));
	} else {
		ut.actime = st.st_atime;
		ut.modtime = start_time;
		utime(file, &ut);
	}
	free(file);
}

static int
bytime(const void *v1, const void *v2)
{
	return ((*(struct item **)v2)->i_time - (*(struct item **)v1)->i_time);
}

static void
allnews(void)
{
	DIR	*Dp;
	struct dirent	*dp;
	struct item	**ip = NULL;
	int	c = 0, i;

	if ((Dp = opendir(".")) == NULL) {
		cantopen(NULL);
		exit(1);
	}
	while ((dp = readdir(Dp)) != NULL) {
		if ((dp->d_name[0] == '.' && (dp->d_name[1] == '\0' ||
		    (dp->d_name[1] == '.' && dp->d_name[2] == '\0'))) ||
				strcmp(dp->d_name, "core") == 0)
			continue;
		ip = srealloc(ip, (c + 1) * sizeof *ip);
		ip[c] = smalloc(sizeof **ip);
		ip[c]->i_name = strdup(dp->d_name);
		ip[c]->i_time = gettime(NULL, ip[c]->i_name, &ip[c]->i_uid);
		c++;
	}
	if (ip == NULL)
		return;
	qsort(ip, c, sizeof *ip, bytime);
	for (i = 0; i < c && (aflag || ip[i]->i_time > news_time); i++)
		item(ip[i]);
}

int
main(int argc, char **argv)
{
	const char	optstring[] = "ans";
	const char	*home;
	int	i;

	progname = basename(argv[0]);
	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'a':
			aflag = 1;
			break;
		case 'n':
			nflag = 1;
			break;
		case 's':
			sflag = 1;
			break;
		default:
			usage();
		}
	}
	if (aflag + nflag + sflag > 1) {
		fprintf(stderr, "%s: options are mutually exclusive\n",
				progname);
		exit(2);
	}
	if ((aflag || nflag || sflag) && argv[optind]) {
		fprintf(stderr, "%s: options are not allowed with file names\n",
				progname);
		exit(2);
	}
	if (chdir(newsdir) < 0) {
		fprintf(stderr, "Cannot chdir to %s\n", newsdir);
		exit(1);
	}
	if ((home = getenv("HOME")) == NULL) {
		fprintf(stderr, "Cannot find HOME variable\n");
		exit(1);
	}
	output = ob_alloc(1, OB_EBF);
	if (!aflag && !nflag && !sflag) {
		if (sigset(SIGINT, SIG_IGN) != SIG_IGN)
			sigset(SIGINT, onint);
	}
	news_time = gettime(home, currency, NULL);
	if (optind != argc) {
		struct item	it;

		for (i = optind; i < argc; i++) {
			it.i_name = argv[i];
			it.i_time = gettime(NULL, it.i_name, &it.i_uid);
			item(&it);
		}
	} else {
		time(&start_time);
		allnews();
		if (!aflag && !nflag && !sflag && item_count > 0)
			touch(home, currency);
	}
	if (nflag) {
		if (item_count > 0)
			printf("\n");
	} else if (sflag) {
		if (item_count > 0)
			printf("%lu news item%s.\n", item_count,
					item_count > 1 ? "s" : "");
		else
			printf("No news.\n");
	}
	return errcnt;
}

/*ARGUSED*/
void
writerr(struct oblok *op, int count, int written)
{
}
